### PR TITLE
Mean, Min, Max, Count global aggs for Expressions2

### PR DESCRIPTION
**Pull request by xcharleslin**

### CODE REVIEW

File: src/series/ops/numeric_agg.rs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+                other
+            ))),
+        }
+    }
+    pub fn mean(&self) -> DaftResult<Series> {
+        use crate::array::ops::DaftNumericAggable;
+        use crate::datatypes::DataType::*;
+
+        match self.data_type() {
+            Int8 | Int16 | Int32 | Int64 => {
```
Comment (samster25) [inline src/series/ops/numeric_agg.rs:39]: This looks copy-pasted from above. Shouldn't all these values be casts to a Float64 and compute a mean on that

File: src/array/ops/count.rs [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+impl<T> DaftCountAggable for &DataArray<T>
+where
+    T: DaftDataType,
+{
+    type Output = DaftResult<DataArray<UInt64Type>>;
+
+    fn count(&self) -> Self::Output {
+        let arrow_array = &self.data;
+        let result_arrow_array = match arrow_array.len() {
+            0 => arrow2::array::PrimitiveArray::from([]),
```
Comment (samster25) [inline src/array/ops/count.rs:18]: I believe this is different compared to pandas and pyarrow which return 0 for an empty array, which i think also makes more sense

File: src/array/ops/numeric_agg.rs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    ($T:ident) => {
+        impl DaftNumericAggable for &DataArray<$T> {
+            type SumOutput = DaftResult<DataArray<$T>>;
+            type MeanOutput = DaftResult<DataArray<Float64Type>>;
+
+            fn sum(&self) -> Self::SumOutput {
+                let primitive_arr = self.downcast();
+
+                let arrow_array = match primitive_arr.len() {
+                    0 => arrow2::array::PrimitiveArray::from([]),
```
Comment (samster25) [inline src/array/ops/numeric_agg.rs:19]: for an empty array, shouldn't be return an array length 1 with a Null

File: src/array/ops/numeric_agg.rs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+                    }
+                };
+                DataArray::new(self.field.clone(), Arc::new(arrow_array))
+            }
+
+            fn mean(&self) -> Self::MeanOutput {
+                let primitive_arr = self.downcast();
+
+                let arrow_array = match primitive_arr.len() {
+                    0 => arrow2::array::PrimitiveArray::from([]),
```
Comment (samster25) [inline src/array/ops/numeric_agg.rs:32]: for an empty array, shouldn't be return an array length 1 with a Null

File: tests/table/test_table.py [hunk 5]
```diff
@@ -274,6 +276,148 @@ def test_table_take_null(idx_dtype) -> None:
     assert taken.to_pydict() == {"a": [None, None], "b": [None, None]}
 
 
+test_table_count_cases = [
+    ([], {"count": []}),
```
Comment (samster25) [inline tests/table/test_table.py:280]: as discussed above, an empty array should be len 0

File: tests/table/test_table.py [hunk 6]
```diff
@@ -274,6 +276,148 @@ def test_table_take_null(idx_dtype) -> None:
...
+    daft_table = Table.from_pydict({"input": [None] * length})
+    daft_table = daft_table.eval_expression_list([col("input").cast(DataType.null())])
+    daft_table = daft_table.eval_expression_list([col("input").alias("count")._count()])
+
+    res = daft_table.to_pydict()["count"]
+    assert res == [length]
+
+
+test_table_minmax_numerics_cases = [
+    ([], {"min": [], "max": []}),
```
Comment (samster25) [inline tests/table/test_table.py:314]: these should yield a None for the min / max

File: tests/table/test_table.py [hunk 7]
```diff
@@ -274,6 +276,148 @@ def test_table_take_null(idx_dtype) -> None:
...
+            col("input").alias("max")._max(),
+        ]
+    )
+
+    res = daft_table.to_pydict()
+    assert res == expected
+
+
+test_table_sum_mean_cases = [
+    ([], {"sum": [], "mean": []}),
```
Comment (samster25) [inline tests/table/test_table.py:395]: these should yield a None for an empty sum / mean

File: src/series/ops/numeric_agg.rs (hunk 1 above)
Comment (samster25) [reply src/series/ops/numeric_agg.rs:39]: if you want to keep it in the trait, we could perform the cast for `DataNumericType` to `Float64` in the impl for `DaftNumericAgg`. Then you can do just dispatch with the macro here.

File: src/array/ops/count.rs (hunk 2 above)
Comment (samster25) [reply src/array/ops/count.rs:18]: aww good catch! I think we changed the behaviors of the nulls in aggregations then. @jaychia Why did we deviate from what pyarrow and pandas does?

Comment (jaychia) [reply src/array/ops/count.rs:18]: We decided on this after drawing up that matrix of behavior across empty/nulls/all-nulls. I don't think we checked that against pyarrow/pandas/postgres behavior actually.

Seems like PostgreSQL does return one empty row for global aggs vs no rows for a group-by agg on an empty table though. We can just follow the same behavior then and change our tests

File: src/dsl/expr.rs [hunk 8]
```diff
@@ -72,6 +87,34 @@ impl AggExpr {
                     },
                 ))
             }
+            Mean(expr) => {
+                let field = expr.to_field(schema)?;
+                Ok(Field::new(
+                    field.name.as_str(),
+                    match &field.dtype {
```
Comment (jaychia) [inline src/dsl/expr.rs:94]: We have a `field.dtype.is_numeric()` now that may be helpful here

**Pull request merged**